### PR TITLE
build: allow unbundling of Node.js dependencies

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Closes https://github.com/electron/electron/issues/44691.

Linux distributions have guidelines on using distro-provided dependencies, rather than compiling them in statically.

This PR this enables downstream packagers to unbundle these dependencies. We don't need to do this for zlib, as the existing gn workflow uses the same `//third_party/zlib`, so unbundling `zlib` with Chromium tools in `//build/linux/unbundle` works already. This adds support for some of the others.

Upstreamed at https://github.com/nodejs/node/pull/55903

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: patches/node/build_allow_unbundling_of_node_js_dependencies.patch [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+diff --git a/node.gni b/node.gni
+index 18d58591e3d0f1f3512db00033c3410a65702864..156cc6e4758fb735368144411c23a853a4ddb197 100644
+--- a/node.gni
++++ b/node.gni
+@@ -60,7 +60,13 @@ declare_args() {
+   # TODO(zcbenz): There are few broken things for now:
+   #   1. cross-os compilation is not supported.
+   #   2. node_mksnapshot crashes when cross-compiling for x64 from arm64.
+-  node_use_node_snapshot = false
++  node_use_node_snapshot = (host_os == target_os) && !(host_cpu == "arm64" && target_cpu == "x64")
```
Comment (deepak1556) [inline patches/node/build_allow_unbundling_of_node_js_dependencies.patch:25]: unrelated change ?

**Review by deepak1556 — APPROVED** [review]

**Review by ckerr — APPROVED** [review]

**Review by jkleinsc — APPROVED** [review]

File: patches/node/build_allow_unbundling_of_node_js_dependencies.patch [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+       deps += [ "//third_party/icu" ]
+     }
+@@ -205,6 +214,19 @@ template("node_gn_build") {
+       sources += node_inspector.node_inspector_sources +
+                  node_inspector.node_inspector_generated_sources
+     }
++    if (is_linux) {
++      import("//build/config/linux/pkg_config.gni")
++        if (use_system_cares) {
++          pkg_config("cares") {
```
Comment (kxxt) [inline patches/node/build_allow_unbundling_of_node_js_dependencies.patch:71]: Nested template invocation causes an error:

```
ERROR at //third_party/electron_node/unofficial.gni:228:11: Can't nest these things.
          pkg_config("cares") {
          ^--------------------
You are trying to nest a template invocation inside a template invocation.
See //third_party/electron_node/unofficial.gni:139:3: The enclosing block.
  source_set("libnode") {
  ^----------------------
See //third_party/electron_node/BUILD.gn:13:1: whence it was called.
node_gn_build("node") {
^----------------------
See //electron/BUILD.gn:246:5: which caused the file to be included.
    "//third_party/electron_node:node_js2c($electron_js2c_toolchain)",
    ^----------------------------------------------------------------
```
